### PR TITLE
Data Pipeline V2: Refactor Deserializer into Serve Input

### LINKED ISSUE #916: Refactor `Deserializer` to be a `DataSource` / `Input`
The deserializers should just be viewed as a `ServeInput`, perhaps with an empty `load_data` and the addition of the sample input property that our sanity checking uses.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #916

This change is temporary until we get merge the ServeInput and add `serve_load_sample` directly to the Input base class.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/core/data/io/input_base.py [hunk 1]
```diff
@@ -232,3 +233,26 @@ def __iter__(self):
...
+    ) -> None:
+        if hasattr(self, "serve_load_data"):
+            raise MisconfigurationException("`serve_load_data` shouldn't be implemented.")
+
+        super().__init__(RunningStage.SERVING, data_pipeline_state=data_pipeline_state)
+
+    def serve_load_sample(self, sample: Any) -> List[Any]:
+        raise NotImplementedError
+
+    def deserialize(self, sample: Any) -> Any:
```
Comment (ethanwharris) [inline flash/core/data/io/input_base.py:251]: Should we instead have this use `__call__`?

File: flash/core/model.py [hunk 2]
```diff
@@ -809,7 +809,13 @@ def build_data_pipeline(
         if deserializer is None or type(deserializer) is Deserializer:
             deserializer = getattr(input_transform, "deserializer", deserializer)
 
-        data_pipeline = DataPipeline(input, input_transform, output_transform, deserializer, output)
+        data_pipeline = DataPipeline(
+            input=input,
+            input_transform=input_transform,
+            output_transform=output_transform,
+            deserializer=deserializer,
```
Comment (ethanwharris) [inline flash/core/model.py:816]: Might be good to refer to it as `serve_input` or something instead of `deserializer`

File: flash/image/data.py [hunk 3]
```diff
@@ -60,7 +60,7 @@ def image_loader(filepath: str):
 
 class ImageDeserializer(Deserializer):
```
Comment (ethanwharris) [inline flash/image/data.py:61]: Maybe this naming would be better:
```suggestion
class ImageServeInput(ServeInput):
```

File: flash/core/data/process.py [hunk 4]
```diff
@@ -12,30 +12,14 @@
...
-    @property
-    @abstractmethod
-    def example_input(self) -> str:
-        raise NotImplementedError
-
-    def __call__(self, sample: Any) -> Any:
-        return self.deserialize(sample)
 
 
 class DeserializerMapping(Deserializer):
```
Comment (ethanwharris) [inline flash/core/data/process.py:25]: We should probably delete `DeserializerMapping`, we don't have a use case for it right now, can always bring it back later if needed

**Review by ethanwharris — COMMENTED** [review]
Looks good :smiley: I guess the existing deserializers should be renamed to end in `ServeInput` and will also need to extend serve input rather than deserializer

File: flash/image/data.py (hunk 3 above)
Comment (tchaton) [reply flash/image/data.py:61]: Let's do this when we get rid of the Deserializer type entirely.

File: flash/core/data/io/input_base.py (hunk 1 above)
Comment (tchaton) [reply flash/core/data/io/input_base.py:251]: Done !

File: flash/core/model.py (hunk 2 above)
Comment (tchaton) [reply flash/core/model.py:816]: I think we would do this with the DataPipeline API change.

File: flash/core/data/process.py (hunk 4 above)
Comment (tchaton) [reply flash/core/data/process.py:25]: This would be done in another PR

**Pull request merged**